virtme/commands/run.py

```diff
@@ -997,43 +1020,89 @@ def cleanup_console_script():
     else:
         os.execvp("ssh", cmd)
 
 
-def ssh_server(args, arch, qemuargs, kernelargs):
-    # Check if we need to generate the ssh host keys for the guest.
-    ssh_key_dir = f"{CACHE_DIR}/.ssh"
-    os.makedirs(f"{ssh_key_dir}/etc/ssh", exist_ok=True)
-    os.system(f"ssh-keygen -A -f {ssh_key_dir}")
-
-    # Implicitly enable dhcp to automatically get an IP on the network
-    # interface and prevent interface renaming.
-    kernelargs.extend(["virtme.dhcp", "net.ifnames=0", "biosdevname=0"])
+def ssh_server(args, arch, qemuargs, kernelargs, *, vsock=True):
+    # Check if we need to generate the SSH host keys for the guest.
+    SSH_ETC_SSH_DIR = SSH_DIR.joinpath("etc", "ssh")
+    SSH_ETC_SSH_DIR.mkdir(mode=0o755, parents=True, exist_ok=True)
+    subprocess.check_call(["ssh-keygen", "-A", "-f", f"{SSH_DIR}"])
 
     # Tell virtme-ng-init / virtme-init to start sshd and use the current
     # username keys/credentials.
     username = get_username()
-    kernelargs.extend(["virtme.ssh"])
-    kernelargs.extend([f"virtme_ssh_user={username}"])
+    if vsock:
+        qemuargs.extend(
+            [
+                "-device",
+                f"{arch.vhost_dev_type('vsock')},guest-cid={args.port}",
+            ]
+        )
+        ssh_channel_type = "vsock"
+    else:
+        # Implicitly enable dhcp to automatically get an IP on the network
+        # interface and prevent interface renaming.
+        kernelargs.extend(["virtme.dhcp", "net.ifnames=0", "biosdevname=0"])
+        # Setup a port forward network interface for the guest.
+        qemuargs.extend(["-device", f"{arch.virtio_dev_type('net')},netdev=ssh"])
+        qemuargs.extend(
+            ["-netdev", f"user,id=ssh,hostfwd=tcp:127.0.0.1:{args.port}-:22"]
+        )
+        ssh_channel_type = "tcp"
+
+    kernelargs.extend(
+        [
+            "virtme.ssh",
+            f"virtme_ssh_channel={ssh_channel_type}",
+            f"virtme_ssh_user={username}",
+        ]
+    )
+    with open(VIRTME_SSH_KNOWN_HOSTS, "w", encoding="utf-8") as f:
+        for path in SSH_ETC_SSH_DIR.glob("*.pub"):
+            pub_key_data = path.open().read()
+            pub_key_data_without_user_and_system = " ".join(
+                pub_key_data.split(" ")[:-1]
+            )
+            f.write(f"localhost {pub_key_data_without_user_and_system}\n")
+            f.write(
+                f"{VIRTME_SSH_DESTINATION_NAME}/* {pub_key_data_without_user_and_system}\n"
+            )
 
-    # Setup a port forward network interface for the guest.
-    qemuargs.extend(["-device", "{},netdev=ssh".format(arch.virtio_dev_type("net"))])
-    qemuargs.extend(["-netdev", f"user,id=ssh,hostfwd=tcp:127.0.0.1:{args.port}-:22"])
+    ssh_proxy = resources.find_script("virtme-ssh-proxy")
+
+    with open(SSH_CONF_FILE, "w", encoding="utf-8") as f:
+        f.write(f"""Host {VIRTME_SSH_DESTINATION_NAME}*
+    CheckHostIP no
+    UserKnownHostsFile {VIRTME_SSH_KNOWN_HOSTS}
```

Comment:
In systemd, they use the following:

```
        # Disable all kinds of host identity checks, since these addresses are generally ephemeral.
        StrictHostKeyChecking no
        UserKnownHostsFile /dev/null
```

https://github.com/poettering/systemd/blob/main/src/ssh-generator/20-systemd-ssh-proxy.conf.in

So maybe fine to do the same, similar to what @arighi suggested? In this case, no need to run `ssh-keygen` above. (this can be done in another PR of course)

Reply:
> In systemd, they use the following:
> 
> ```
>         # Disable all kinds of host identity checks, since these addresses are generally ephemeral.
>         StrictHostKeyChecking no
>         UserKnownHostsFile /dev/null
> ```
> 
> https://github.com/poettering/systemd/blob/main/src/ssh-generator/20-systemd-ssh-proxy.conf.in
> 
> So maybe fine to do the same, similar to what @arighi suggested? In this case, no need to run `ssh-keygen` above. (this can be done in another PR of course)

Yeah, we should probably do that, adding that comment as well to explain why we're doing that in case people are looking at the conf file. I think it can save some potential pain.

Reply:
I agree, "micro optimizations" can be done later :)